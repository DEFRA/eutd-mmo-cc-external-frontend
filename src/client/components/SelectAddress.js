import React from 'react';
import { GridRow, GridCol } from 'govuk-react';
import Select from '../../../node_modules/govuk-react-components/components/forms/select.js';
import SecondaryButton from './elements/SecondaryButton';
import ContinueButton from './elements/ContinueButton';
import { withTranslation } from 'react-i18next';

const steps = {
  NEXT: 'NEXT',
  PREV: 'PREV',
  SKIP: 'SKIP'
};

const SelectAddress = ({
  postcode,
  addresses,
  setSelectedAddress,
  handleCancelButton,
  handleSelectAddressNavigation,
  errors,
  t
}) => {
  const [address, setAddress] = React.useState('');

  const options = addresses && addresses.map(addr => addr.address_line);

  const nullOption = addresses && addresses.length
    ? `${addresses.length} ${addresses.length > 1 ? t('commonWhatExportersAddressMultipleAddressesFound') : t('commonWhatExportersAddressSingleAddressFound')}`
    : t('commonWhatExportersAddressNoAddressesFound');

  const handleChange = (evt) => {
    setAddress(evt.currentTarget.value);
    setSelectedAddress(evt.currentTarget.value);
  };

  return (
    <>
      <GridRow>
        <GridCol>
          <label className="heading-small">{t('commonWhatExportersAddressPostcode')}</label>
          <p>
            {postcode}
            <button
              id="selectAddressNavigationPrev"
              className="button-dressed-as-link"
              onClick={() => handleSelectAddressNavigation('PREV')}
            >
              {t('commonWhatExportersAddressChangeLink')}
            </button>
          </p>
          <span className="prop-dropdown state">
            <Select
              label={t('commonWhatExportersAddressSelectAddress')}
              name="selectAddress"
              id="selectAddress"
              value={address}
              error={t(errors.addressError)}
              defaultValue={address}
              options={options}
              onChange={handleChange}
              nullOption={nullOption}
              key={address}
            />
          </span>
        </GridCol>
      </GridRow>
      <GridRow>
        <SecondaryButton
          id="cancelSelectAddress"
          name="cancelSelectAddress"
          value="cancel"
          onClick={handleCancelButton}
        >
          {t('commonSecondaryButtonCancelButton')}
        </SecondaryButton>

        <ContinueButton
          id="continueSelectAddress"
          name="continueSelectAddress"
          label="Continue"
          onClick={() => handleSelectAddressNavigation(steps.NEXT)}
        >
          {t('commonContinueButtonContinueButtonText')}
        </ContinueButton>
      </GridRow>
      <GridRow>
        <GridCol>
          <p>
            <button
              id="selectAddressNavigationSkip"
              className="button-dressed-as-link"
              onClick={() => handleSelectAddressNavigation(steps.SKIP)}
            >
              {t('commonWhatExportersAddressAddressNotFoundText')}
            </button>
          </p>
        </GridCol>
      </GridRow>
    </>
  );
};

export default (withTranslation()(SelectAddress));