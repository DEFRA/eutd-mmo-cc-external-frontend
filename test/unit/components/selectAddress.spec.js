import React from 'react';
import { mount } from 'enzyme';

import SelectAddress from '../../../src/client/components/SelectAddress';
import { act } from 'react-dom/test-utils';
import { render } from '@testing-library/react';

describe('Test Suite - Postcode Look up - Step 2: Select Address', () => {
  const props = {
    postcode: 'B20 3RU',
    errors: { error: [], addressError: '' },
    addresses: [
      {
        address_line: 'MARINE MANAGEMENT ORGANIZATION, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH',
        building_name: 'LANCASTER HOUSE',
        street_name: 'HAMPSHIRE COURT',
        county: 'TYNESIDE',
        country: 'ENGLAND',
        city: 'NEWCASTLE UPON TYNE',
        postCode: 'NE4 7YH',
      },
    ],
    setSelectedAddress: jest.fn(),
    handleCancelButton: jest.fn(),
    handleSelectAddressNavigation: jest.fn(),
  };

  const wrapper = mount(<SelectAddress {...props} />);

  it('Should render "Select Address" component', () => {
    expect(wrapper).toBeDefined();
  });

  it('Should render drop-down component', () => {
    const dropdown = wrapper.find('select[name=\'selectAddress\']');
    expect(dropdown.exists()).toBeTruthy();
  });
  it('Should keep a default option selected in drop-down component', () => {
    const dropdown = wrapper.find('select[name=\'selectAddress\']');
    expect(dropdown.childAt(0).text()).toBe('1 address found');
  });

  it('Should have default option text updated according to number of results in drop-down component', () => {
    const props = {
      postcode: 'B20 3RU',
      errors: { error: [], addressError: '' },
      addresses: [
        {
          address_line: 'MARINE MANAGEMENT ORGANIZATION, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH',
          building_name: 'LANCASTER HOUSE',
          street_name: 'HAMPSHIRE COURT',
          county: 'TYNESIDE',
          country: 'ENGLAND',
          city: 'NEWCASTLE UPON TYNE',
          postCode: 'NE4 7YH',
        },
        {
          address_line: 'MARINE MANAGEMENT ORGANIZATION, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH',
          building_name: 'LANCASTER HOUSE',
          street_name: 'HAMPSHIRE COURT',
          county: 'TYNESIDE',
          country: 'ENGLAND',
          city: 'NEWCASTLE UPON TYNE',
          postCode: 'NE4 9YH',
        },
        {
          address_line: 'MARINE MANAGEMENT ORGANIZATION, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH',
          building_name: 'LANCASTER HOUSE',
          street_name: 'HAMPSHIRE COURT',
          county: 'TYNESIDE',
          country: 'ENGLAND',
          city: 'NEWCASTLE UPON TYNE',
          postCode: 'NE4 8YH',
        }
      ],
      setSelectedAddress: jest.fn(),
      handleCancelButton: jest.fn(),
      handleSelectAddressNavigation: jest.fn(),
    };
    const wrapper = mount(<SelectAddress {...props} />);
    const dropdown = wrapper.find('select[name=\'selectAddress\']');
    expect(dropdown.childAt(0).text()).toBe('3 addresses found');
  });

  it('Should have default option text updated to No addresses found when no address were found', () => {
    const props = {
      postcode: 'B20 3RU',
      errors: { error: [], addressError: '' },
      addresses: [],
      setSelectedAddress: jest.fn(),
      handleCancelButton: jest.fn(),
      handleSelectAddressNavigation: jest.fn(),
    };
    const wrapper = mount(<SelectAddress {...props} />);
    const dropdown = wrapper.find('select[name=\'selectAddress\']');
    expect(dropdown.childAt(0).text()).toBe('No addresses found');
  });

  it('Should have an option available for selection by user in drop-down component when only building_name is given', () => {
    const dropdown = wrapper.find('select[name=\'selectAddress\']');
    expect(dropdown.childAt(1).text().trim()).toBe(
      'MARINE MANAGEMENT ORGANIZATION, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH'
    );
  });

  it('should show correct error when there is any', () => {
    const props = {
      postcode: 'B20 3RU',
      errors: {
        errors: [
        {
          targetName: 'postcode', text: 'commonLookupAddressPageErrorSelectedAddress'
        }
        ],
        addressError: 'commonLookupAddressPageErrorSelectedAddress' },
      addresses: [],
      setSelectedAddress: jest.fn(),
      handleCancelButton: jest.fn(),
      handleSelectAddressNavigation: jest.fn(),
    };
    const wrapper = mount(<SelectAddress {...props} />);
    expect(wrapper.find('.error-message').first().text()).toBe('Select an address to continue or click \'I cannot find my address in the list\'');
  });

  it('should take a snapshot of SelectAddress: PostcodeLookUp: Step 2', ()=> {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });
});

describe('select option dropdown list', () => {

  const props = {
    postcode: 'B20 3RU',
    errors: { error: [], addressError: '' },
    addresses: [
      {
        address_line: 'MARINE MANAGEMENT ORGANIZATION 1, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH',
        building_number: '34',
        building_name: 'LANCASTER HOUSE',
        sub_building_name: 'TOP FLOOR',
        street_name: 'HAMPSHIRE COURT',
        county: 'TYNESIDE',
        country: 'ENGLAND',
        city: 'NEWCASTLE UPON TYNE',
        postCode: 'NE4 7YH',
      },
      {
        address_line: 'MARINE MANAGEMENT ORGANIZATION 2, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH',
        building_number: '34',
        street_name: 'HAMPSHIRE COURT',
        county: 'TYNESIDE',
        country: 'ENGLAND',
        city: 'NEWCASTLE UPON TYNE',
        postCode: 'NE4 7YH',
      },
      {
        address_line: 'MARINE MANAGEMENT ORGANIZATION 3, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH',
        building_name: 'LANCASTER HOUSE',
        street_name: 'HAMPSHIRE COURT',
        county: 'TYNESIDE',
        country: 'ENGLAND',
        city: 'NEWCASTLE UPON TYNE',
        postCode: 'NE4 7YH',
      }
    ],
    setSelectedAddress: jest.fn(),
    handleCancelButton: jest.fn(),
    handleSelectAddressNavigation: jest.fn(),
  };

  const wrapper = mount(<SelectAddress {...props} />);

  it('should included an option with all properties when all are given', () => {
    const dropdown = wrapper.find('select[name=\'selectAddress\']');
    expect(dropdown.childAt(1).text().trim()).toBe(
      'MARINE MANAGEMENT ORGANIZATION 1, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH'
    );
  });

  it('should included an option with only building_number and main properties when only building_number is added', () => {
    const dropdown = wrapper.find('select[name=\'selectAddress\']');
    expect(dropdown.childAt(2).text().trim()).toBe(
      'MARINE MANAGEMENT ORGANIZATION 2, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH'
    );
  });
  it('should set address with the value of the select address', () => {
    const dropdown = wrapper.find('select[name=\'selectAddress\']');
    act(() => {
      dropdown.prop('onChange')({ currentTarget: { value: 'blah'}});
    });

    expect(props.setSelectedAddress).toHaveBeenCalled();
    expect(props.setSelectedAddress).toHaveBeenCalledWith('blah');
  });

  it('should call handleSelectAddressNavigation', () => {
    const dropdown = wrapper.find('#selectAddressNavigationPrev');
    act(() => {
      dropdown.prop('onClick')({});
    });

    expect(props.handleSelectAddressNavigation).toHaveBeenCalled();
    expect(props.handleSelectAddressNavigation).toHaveBeenCalledWith('PREV');
  });

  it('should call handleSelectAddressNavigation when continuing', () => {
    const dropdown = wrapper.find('#continueSelectAddress');
    act(() => {
      dropdown.first().prop('onClick')({});
    });

    expect(props.handleSelectAddressNavigation).toHaveBeenCalled();
    expect(props.handleSelectAddressNavigation).toHaveBeenCalledWith('NEXT');
  });

  it('should call handleSelectAddressNavigation when skipping', () => {
    const dropdown = wrapper.find('#selectAddressNavigationSkip');
    act(() => {
      dropdown.prop('onClick')({});
    });

    expect(props.handleSelectAddressNavigation).toHaveBeenCalled();
    expect(props.handleSelectAddressNavigation).toHaveBeenCalledWith('SKIP');
  });
});