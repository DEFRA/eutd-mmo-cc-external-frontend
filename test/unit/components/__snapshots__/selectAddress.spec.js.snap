// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Suite - Postcode Look up - Step 2: Select Address should take a snapshot of SelectAddress: PostcodeLookUp: Step 2 1`] = `
<div>
  <div
    class="css-1lzr9ic-StyledRow e4u1u2g0"
  >
    <div
      class="css-1w80l5h-StyledColumn e4u1u2g0"
    >
      <label
        class="heading-small"
      >
        Postcode
      </label>
      <p>
        B20 3RU
        <button
          class="button-dressed-as-link"
          id="selectAddressNavigationPrev"
        >
          Change
        </button>
      </p>
      <span
        class="prop-dropdown state"
      >
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="selectAddress"
          >
            <span
              class="form-label-bold"
            >
              Select an address
            </span>
            
          </label>
          <select
            class="form-control"
            id="selectAddress"
            name="selectAddress"
          >
            <option
              value=""
            >
              1 address found
            </option>
            <option
              value="MARINE MANAGEMENT ORGANIZATION, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH"
            >
              MARINE MANAGEMENT ORGANIZATION, LANCASTER HOUSE, HAMPSHIRE COURT, NEWCASTLE UPON TYNE, NE4 7YH
            </option>
          </select>
        </div>
      </span>
    </div>
  </div>
  <div
    class="css-1lzr9ic-StyledRow e4u1u2g0"
  >
    <button
      class="css-17uw1q2-StyledButton-StyledHoc e4u1u2g0"
      id="cancelSelectAddress"
      name="cancelSelectAddress"
      value="cancel"
    >
      Cancel
    </button>
    <button
      class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
      id="continueSelectAddress"
      label="Continue"
      name="continueSelectAddress"
    >
      Continue
    </button>
  </div>
  <div
    class="css-1lzr9ic-StyledRow e4u1u2g0"
  >
    <div
      class="css-1w80l5h-StyledColumn e4u1u2g0"
    >
      <p>
        <button
          class="button-dressed-as-link"
          id="selectAddressNavigationSkip"
        >
          I can not find the address in the list
        </button>
      </p>
    </div>
  </div>
</div>
`;
